README.md
# MineArea Minecraft Modpack

This repository contains a few files for creating the zip files for the MineArea Minecraft modpack.

## How to build

* Download the [Minecraft Forge installer](https://files.minecraftforge.net) and put it into `source/forge`
* Install and update mods (see section bellow)
* Build the modpack by executing `make` in the root directory of this repository
* Optional: Build the Docker image for the Minecraft server using `make server-docker`

## Install and update mods

Execute `download-mods.py` to download all mods listed in the mods.yaml file.

In case you want to update the mods, specify the `--update` option: `download-mods.py --update`

**Note:** This currently only works for mods available on CurseForge.

Don't forget to update the [change log](CHANGES.md) after updating the mods (use the information from the updated-mods.txt file).

CHANGES.md
# Changes

## 1.14 (2019-07-31)

Updated mods:
- [Applied Energistics 2](https://www.curseforge.com/minecraft/mc-mods/applied-energistics-2): rv6-stable-6 -> rv6-stable-7
- [Carpenter's Blocks](https://www.carpentersblocks.com): v3.4.0-poc.5 -> v3.4.0-poc.6
- [Ender Core](https://www.curseforge.com/minecraft/mc-mods/endercore): 0.5.57 -> 0.5.59
- [Ender IO](https://www.curseforge.com/minecraft/mc-mods/ender-io): 5.0.44 -> 5.0.46
- [Extreme Reactors](https://www.curseforge.com/minecraft/mc-mods/extreme-reactors): 0.4.5.66 -> 0.4.5.67
- [In Control](https://www.curseforge.com/minecraft/mc-mods/in-control): 3.9.10 -> 3.9.13
- [Industrial Foregoing](https://www.curseforge.com/minecraft/mc-mods/industrial-foregoing): 1.12.12 -> 1.12.13
- [Just Enough Energistics (JEE)](https://www.curseforge.com/minecraft/mc-mods/just-enough-energistics-jee): 1.0.6 -> 1.0.7
- [Pam's HarvestCraft](https://www.curseforge.com/minecraft/mc-mods/pams-harvestcraft): 1.12.2zd -> 1.12.2ze
- [Tinkers Construct](https://www.curseforge.com/minecraft/mc-mods/tinkers-construct): 2.12.0.135 -> 2.12.0.157
- [Underground Biomes](https://www.curseforge.com/minecraft/mc-mods/underground-biomes): 1.3.5 -> 1.3.6

## 1.13 (2019-06-08)

New mods:
- [Chicken Chunks](https://www.curseforge.com/minecraft/mc-mods/chicken-chunks-1-8)

Updated mods:
- [Backpacks!](https://www.curseforge.com/minecraft/mc-mods/backpacks): 3.5.5 -> 3.5.6
- [Biomes O Plenty](https://www.curseforge.com/minecraft/mc-mods/biomes-o-plenty): 7.0.1.2427 -> 7.0.1.2439
- [Chisel & Bits](https://www.curseforge.com/minecraft/mc-mods/chisels-bits): 14.32 -> 14.33
- [CoFH Core](https://www.curseforge.com/minecraft/mc-mods/cofh-core): 4.6.2.25 -> 4.6.3.27
- [CoFH World](https://www.curseforge.com/minecraft/mc-mods/cofh-world): 1.3.0.6 -> 1.3.1.7
- [In Control](https://www.curseforge.com/minecraft/mc-mods/in-control): 3.9.7 -> 3.9.10
- [Mantle](https://www.curseforge.com/minecraft/mc-mods/mantle): 1.3.3.49 -> 1.3.3.55
- [MrCrayfish Furniture Mod](https://mrcrayfish.com/mods?id=cfm): 6.2.0 -> 6.3.0
- [Pam's HarvestCraft](https://www.curseforge.com/minecraft/mc-mods/pams-harvestcraft): 1.12.2zc -> 1.12.2zd
- [Thermal Dynamics](https://www.curseforge.com/minecraft/mc-mods/thermal-dynamics): 2.5.4.18 -> 2.5.5.21
- [Thermal Expansion](https://www.curseforge.com/minecraft/mc-mods/thermal-expansion): 5.5.3.41 -> 5.5.4.43
- [Thermal Foundation](https://www.curseforge.com/minecraft/mc-mods/thermal-foundation): 2.6.2.26 -> 2.6.3.27
- [Underground Biomes](https://www.curseforge.com/minecraft/mc-mods/underground-biomes): 1.3.4 -> 1.3.5

## 1.12 (2019-05-13)

Updated Forge to version 14.23.5.2837

New mods:
- [Crafting Tweaks](https://www.curseforge.com/minecraft/mc-mods/crafting-tweaks)
- [Ping](https://www.curseforge.com/minecraft/mc-mods/ping)
- [TorchMaster](https://www.curseforge.com/minecraft/mc-mods/torchmaster)
- [UniDict](https://www.curseforge.com/minecraft/mc-mods/unidict)

Updated mods:
- [Ambient Sounds](https://www.curseforge.com/minecraft/mc-mods/ambientsounds): 3.0.5 -> 3.0.9
- [Chisel & Bits](https://www.curseforge.com/minecraft/mc-mods/chisels-bits): 14.31 -> 14.32
- [Ender IO](https://www.curseforge.com/minecraft/mc-mods/ender-io): 5.0.43 -> 5.0.44
- [Extreme Reactors](https://www.curseforge.com/minecraft/mc-mods/extreme-reactors): 0.4.5.65 -> 0.4.5.66
- [Hammer (Lib) Core](https://www.curseforge.com/minecraft/mc-mods/hammer-lib): 2.0.4.4 -> 2.0.4.7
- [In Control](https://www.curseforge.com/minecraft/mc-mods/in-control): 3.9.6 -> 3.9.7
- [Industrial Foregoing](https://www.curseforge.com/minecraft/mc-mods/industrial-foregoing): 1.12.11 -> 1.12.12
- [NoMoreRecipeConflict](https://www.curseforge.com/minecraft/mc-mods/stimmedcow-nomorerecipeconflict): 0.12 -> 0.13
- [Underground Biomes](https://www.curseforge.com/minecraft/mc-mods/underground-biomes): 1.3.3 -> 1.3.4
- [Wawla](https://www.curseforge.com/minecraft/mc-mods/wawla-what-are-we-looking-at): 2.5.269 -> 2.5.273

## 1.11 (2019-05-02)

New mods:
- [Just Enough Energistics (JEE)](https://www.curseforge.com/minecraft/mc-mods/just-enough-energistics-jee)

## 1.10 (2019-04-28)

New mods:
- [ExtraCells2](https://www.curseforge.com/minecraft/mc-mods/extracells2)

## 1.9 (2019-04-22)

Updated Forge to version 14.23.5.2836

Updated mods:
- [Ambient Sounds](https://www.curseforge.com/minecraft/mc-mods/ambientsounds): 3.0.3 -> 3.0.5
- [Biomes O Plenty](https://www.curseforge.com/minecraft/mc-mods/biomes-o-plenty): 7.0.1.2425 -> 7.0.1.2427
- [BuildCraft](https://www.curseforge.com/minecraft/mc-mods/buildcraft): 7.99.23 -> 7.99.24.1
- [EnderCore](https://www.curseforge.com/minecraft/mc-mods/endercore): 0.5.52 -> 0.5.57
- [Forestry](https://www.curseforge.com/minecraft/mc-mods/forestry): 5.8.2.382 -> 5.8.2.387
- [InControl](https://www.curseforge.com/minecraft/mc-mods/in-control): 3.8.3 -> 3.9.6
- [OpenBlocks](https://www.curseforge.com/minecraft/mc-mods/openblocks): 1.8 -> 1.8.1
- [OpenModsLib](https://www.curseforge.com/minecraft/mc-mods/openmodslib): 0.12.1 -> 0.12.2
- [Pam's HarvestCraft](https://www.curseforge.com/minecraft/mc-mods/pams-harvestcraft): 1.12.2zb -> 1.12.2zc
- [Plants](https://www.curseforge.com/minecraft/mc-mods/plants): 2.10.6 -> 2.10.7
- [Underground Biomes](https://www.curseforge.com/minecraft/mc-mods/underground-biomes): 1.3.2 -> 1.3.3

Removed mods:
- [Better Foliage](https://www.curseforge.com/minecraft/mc-mods/better-foliage) (huge performance decrease!)
- [Railcraft](https://www.curseforge.com/minecraft/mc-mods/railcraft) (not used on our server)

## 1.8 (2019-04-06)

New mods:
- [AutoPlant](https://www.curseforge.com/minecraft/mc-mods/autoplant)
- [Better Foliage](https://www.curseforge.com/minecraft/mc-mods/better-foliage)
- [MrCrayfish Furniture Mod](https://mrcrayfish.com/mods?id=cfm)
- [OreDictionaryConverter](https://www.curseforge.com/minecraft/mc-mods/oredictionaryconverter)
- Terra
- [Underground Biomes](https://www.curseforge.com/minecraft/mc-mods/underground-biomes)

## 1.7 (2019-04-05)

docker/mcrcon.py
#! /usr/bin/env python3

import os
import readline
import sys

from mcrcon import MCRcon
from jproperties import Properties

server_roperties_file = os.path.join(os.getenv("MINECRAFT_DIR"), "server.properties")
history_file = os.path.expanduser("~/.mcrcon_history")

if not os.path.exists(server_roperties_file):
    print("{} not found!".format(server_roperties_file))
    exit(1)

server_properties = Properties.load(server_roperties_file)

if "enable-rcon" not in server_properties or server_properties["enable-rcon"].lower() != "true":
    print("RCON is not enabled!")
    exit(1)

if "rcon.port" not in server_properties:
    print("RCON port is not configured!")
    exit(1)

if "rcon.password" not in server_properties:
    print("RCON password is not configured!")
    exit(1)

rcon_host = "localhost"
rcon_port = int(server_properties["rcon.port"])
rcon_password = server_properties["rcon.password"]

with MCRcon(host=rcon_host, port=rcon_port, password=rcon_password) as rcon:
    if len(sys.argv) < 2:
        if os.isatty(sys.stdout.fileno()):
            if os.path.isfile(history_file):
                readline.read_history_file(history_file)

            try:
                while True:
                    command = input("{}:{} > ".format(rcon_host, rcon_port)).strip()

                    if not command:
                        continue

                    command_lower = command.lower()
                    if command_lower == "quit" or command_lower == "exit":
                        break

                    response = rcon.command(command)
                    if response:
                        print(response, end="\n")
            except (EOFError, KeyboardInterrupt):
                pass
            finally:
                readline.write_history_file(history_file)

                rcon.disconnect()
        else:
            print("Usage: {} <command>".format(sys.argv[0]), file=sys.stderr)
            exit(1)
    else:
        print(rcon.command(" ".join(sys.argv[1:])))


update-mc-mods.py
#! /usr/bin/env python3

import argparse
import glob
import hashlib
import json
import os
import re
import requests
import sys
import yaml
import zipfile

from urllib.parse import urlparse


class MCModInfo:
    def __init__(self, jar_filename, pattern):
        self.version = None

        if jar_filename is None:
            return

        try:
            zip_file = zipfile.ZipFile(jar_filename)
            mod_info = json.loads(zip_file.read("mcmod.info"))

            mod_info = mod_info[0]
        except Exception:
            mod_info = {}

        if "version" in mod_info:
            self.version = mod_info["version"]
        else:
            # Try to get version from file pattern if it contains a wildcard character (*)
            if "*" in pattern:
                match = re.match(pattern.replace("*", "(.*)"), os.path.basename(jar_filename))
                if match:
                    self.version = match.group(1)

            # Use filename without .jar extension if version is still None
            if self.version is None:
                self.version = os.path.splitext(os.path.basename(jar_filename))[0]


class Mod:
    def __init__(self, mods_dir, data):
        self.name: str = data["name"]
        self.pattern: str = data["pattern"]
        self.url: str = data["url"]
        self.download_url: str = None
        self.latest_filename: str = None
        self.md5: str = None

        if "download_url" in data and data["download_url"] is not None:
            self.download_url = data["download_url"]
            self.latest_filename = os.path.basename(self.download_url)

        if "md5" in data:
            self.md5 = data["md5"]

        matching_files = glob.glob(os.path.join(mods_dir, self.pattern))

        if len(matching_files) > 1:
            raise RuntimeError("More than one file matches pattern {}".format(self.pattern))

        if matching_files:
            self.filename = matching_files[0]
        else:
            self.filename = None

    def to_yaml(self):
        yaml_data = {
            "name": self.name,
            "pattern": self.pattern,
            "url": self.url,
            "download_url": self.download_url,
            "md5": self.md5
        }

        return yaml_data

    def check_url(self):
        parsed_url = urlparse(self.url)

        if parsed_url.netloc != "minecraft.curseforge.com":
            print("Skipping unsupported URL: {}".format(self.url), file=sys.stderr)
            return False

        return True

    def get_modinfo(self):
        return MCModInfo(self.filename, self.pattern)

    def update_latest_file(self):
        response = requests.get("{}/files/latest".format(self.url), allow_redirects=False)

        response.raise_for_status()

        self.download_url = response.headers["Location"]
        self.latest_filename = os.path.basename(self.download_url)

    def download(self, filename):
        print("Downloading {} to {}".format(self.download_url, filename))

        try:
            with requests.get(self.download_url, stream=True) as response:
                response.raise_for_status()
                with open(filename, "wb") as local_file:
                    for chunk in response.iter_content(chunk_size=8192):
                        if chunk:  # filter out keep-alive new chunks
                            local_file.write(chunk)
            return True
        except Exception as exception:
            print("Download failed: {}".format(exception), file=sys.stderr)
            return False

    def is_file_valid(self):
        if self.filename is None:
            return False

        if self.md5 is not None and md5file(self.filename) != self.md5:
            return None

        return True


def md5file(filename):
    md5 = hashlib.md5()

    with open(filename, "rb") as file:
        while True:
            data = file.read(65536)

            if not data:
                break

            md5.update(data)

    return md5.hexdigest()


def main():
    argument_parser = argparse.ArgumentParser(description="MineArea mods updater")

    argument_parser.add_argument("--no-update", help="only download new or missing mods, do not update existing mods", action="store_true")

    cmd_arguments = argument_parser.parse_args()

    root_dir = os.path.dirname(os.path.realpath(__file__))
    mods_dir = os.path.join(root_dir, "source", "mods")

    if not os.path.exists(mods_dir):
        os.mkdir(mods_dir)

    mod_list_file = os.path.join(root_dir, "mods.yaml")
    update_info_file = os.path.join(root_dir, "updated-mods.txt")

    update_info_data = {
        "new": [],
        "update": []
    }

    with open(mod_list_file) as mod_list_yaml:
        mods = yaml.safe_load(mod_list_yaml)

        mods = [Mod(mods_dir, mod) for mod in mods]

        mods_with_update = []

        for mod in mods:
            if cmd_arguments.no_update and mod.download_url is not None and not mod.is_file_valid():
                mods_with_update.append(mod)
                continue

            if not mod.check_url():
                continue

            if cmd_arguments.no_update and mod.filename is not None:
                continue

            mod.update_latest_file()

            if mod.filename is None:
                print("New mod {} found: {}".format(mod.name, mod.latest_filename))
            else:
                current_filename = os.path.basename(mod.filename)

                if mod.latest_filename == current_filename:
                    print("No update found for {}".format(mod.name), file=sys.stderr)
                    continue

                print("Update for {} found: {} -> {}".format(mod.name, current_filename, mod.latest_filename))

            mods_with_update.append(mod)

        if not mods_with_update:
            print("Nothing to update")
            return

        print("Ready to download {} mods".format(len(mods_with_update)))
        input("Press Enter to continue...")

        for mod in mods_with_update:
            download_filepath = os.path.join(mods_dir, mod.latest_filename)
            if mod.download(download_filepath):
                old_version = mod.get_modinfo().version
                new_version = MCModInfo(download_filepath, mod.pattern).version

                if mod.filename is None:
                    update_info_data["new"].append("[{}]({})".format(mod.name, mod.url))
                elif not cmd_arguments.no_update:
                    update_info_data["update"].append("[{}]({}): {} -> {}".format(mod.name, mod.url, old_version, new_version))

                # Remove old mod file
                if mod.filename is not None and os.path.exists(mod.filename):
                    os.remove(mod.filename)

                md5 = md5file(download_filepath)

                if cmd_arguments.no_update:
                    if mod.md5 is not None and mod.md5 != md5:
                        print("Warning: MD5 checksum of downloaded file does not match stored MD5 checksum!")
                else:
                    mod.md5 = md5
            elif os.path.exists(download_filepath):
                os.remove(download_filepath)

    # Dump any changes to the YAML structure (e.g. new download_url) back to the mods.yaml file
    with open(mod_list_file, "w") as mod_list_yaml:
        yaml_list = []

        for mod in mods:
            yaml_list.append(mod.to_yaml())

        yaml.dump(yaml_list, mod_list_yaml, default_flow_style=False, sort_keys=False)

    with open(update_info_file, "w") as update_info:
        if update_info_data["new"]:
            update_info.write("New mods:\n")
            for line in update_info_data["new"]:
                update_info.write("- {}\n".format(line))

        if update_info_data["update"]:
            if update_info_data["new"]:
                update_info.write("\n")

            update_info.write("Updated mods:\n")
            for line in update_info_data["update"]:
                update_info.write("- {}\n".format(line))

    print("Done")


if __name__ == "__main__":
    main()


docker/entrypoint.sh
#! /bin/bash

rsync -a --exclude server.properties /usr/src/minecraft/ ${MINECRAFT_DIR}/

rsync -a --delete \
      /usr/src/minecraft/libraries \
      /usr/src/minecraft/mods \
      ${MINECRAFT_DIR}

if [[ ! -e ${MINECRAFT_DIR}/server.properties ]]; then
    cp /usr/src/minecraft/server.properties ${MINECRAFT_DIR}/server.properties
fi

find ${MINECRAFT_DIR} \! -user ${MINECRAFT_USER} -exec chown ${MINECRAFT_USER}:${MINECRAFT_GROUP} '{}' +

case "$1" in
    server)
        java_opts=("java")

        if [[ -z ${JAVA_HEAP_MAX} ]]; then
            java_opts+=("-XX:+UnlockExperimentalVMOptions" "-XX:+UseCGroupMemoryLimitForHeap" "-XX:MaxRAMFraction=1")
        else
            java_opts+=("-Xmx${JAVA_HEAP_MAX}")

            if [[ ! -z ${JAVA_HEAP_MIN} ]]; then
                java_opts+=("-Xms${JAVA_HEAP_MIN}")
            fi
        fi

        if [[ ${JMXREMOTE_ENABLE} ]]; then
            java_opts+=("-Dcom.sun.management.jmxremote" "-Dcom.sun.management.jmxremote.port=${JMXREMOTE_PORT}" "-Dcom.sun.management.jmxremote.rmi.port=${JMXREMOTE_PORT}" "-Dcom.sun.management.jmxremote.local.only=false" "-Dcom.sun.management.jmxremote.authenticate=false" "-Dcom.sun.management.jmxremote.ssl=false" "-Djava.rmi.server.hostname=${JVMREMOTE_HOSTNAME}")
        fi

        if [[ ! -z ${JAVA_OPTS} ]]; then
            java_opts+=(${JAVA_OPTS})
        fi

        java_opts+=("-jar" "${MINECRAFT_DIR}/minecraftforge.jar" "--noconsole")

        if ! fgrep -qs "enable-rcon" ${MINECRAFT_DIR}/server.properties; then
            echo "enable-rcon=true" >> ${MINECRAFT_DIR}/server.properties
        fi

        if ! fgrep -qs "rcon.port" ${MINECRAFT_DIR}/server.properties; then
            echo "rcon.port=25575" >> ${MINECRAFT_DIR}/server.properties
        fi

        if ! fgrep -qs "rcon.password" ${MINECRAFT_DIR}/server.properties; then
            echo "rcon.password=$(date +%s | sha256sum | base64 | head -c 32)" >> ${MINECRAFT_DIR}/server.properties
        fi

        echo "Starting Minecraft server: ${java_opts[@]}"

        exec gosu ${MINECRAFT_USER} "${java_opts[@]}"
    ;;

    *)
        exec "$@"
    ;;
esac

download-mods.py
#! /usr/bin/env python3

import argparse
import glob
import hashlib
import json
import os
import re
import requests
import sys
import yaml
import zipfile

from urllib.parse import urlparse, unquote


class MCModInfo:
    def __init__(self, jar_filename, pattern):
        self.version = None

        if jar_filename is None:
            return

        try:
            zip_file = zipfile.ZipFile(jar_filename)
            mod_info = json.loads(zip_file.read("mcmod.info"))

            mod_info = mod_info[0]
        except Exception:
            mod_info = {}

        if "version" in mod_info:
            self.version = mod_info["version"]
        else:
            # Try to get version from file pattern if it contains a wildcard character (*)
            if "*" in pattern:
                match = re.match(pattern.replace("*", "(.*)"), os.path.basename(jar_filename))
                if match:
                    self.version = match.group(1)

            # Use filename without .jar extension if version is still None
            if self.version is None:
                self.version = os.path.splitext(os.path.basename(jar_filename))[0]


class Mod:
    def __init__(self, mods_dir, data):
        self.name: str = data["name"]
        self.pattern: str = data["pattern"]
        self.url: str = data["url"]
        self.download_url: str = None
        self.latest_filename: str = None
        self.md5: str = None

        if "download_url" in data and data["download_url"] is not None:
            self.download_url = data["download_url"]
            self.set_latest_file_from_download_url()

        if "md5" in data:
            self.md5 = data["md5"]

        matching_files = glob.glob(os.path.join(mods_dir, self.pattern))

        if len(matching_files) > 1:
            raise RuntimeError("More than one file matches pattern {}".format(self.pattern))

        if matching_files:
            self.filename = matching_files[0]
        else:
            self.filename = None

    def to_yaml(self):
        yaml_data = {
            "name": self.name,
            "pattern": self.pattern,
            "url": self.url,
            "download_url": self.download_url,
            "md5": self.md5
        }

        return yaml_data

    def check_url(self):
        parsed_url = urlparse(self.url)

        if parsed_url.netloc != "www.curseforge.com":
            print("Skipping unsupported URL: {}".format(self.url), file=sys.stderr)
            return False

        return True

    def get_modinfo(self):
        return MCModInfo(self.filename, self.pattern)

    def set_latest_file_from_download_url(self):
        self.latest_filename = os.path.basename(unquote(urlparse(self.download_url).path).replace("+", " "))

    def update_latest_file(self):
        response = requests.get("{}/download".format(self.url), allow_redirects=False)

        response.raise_for_status()

        match = re.search("/minecraft/mc-mods/[^/]+/download/([0-9]+)/file", response.text)

        response = requests.get("https://www.curseforge.com{}".format(match.group(0)), allow_redirects=False)

        response.raise_for_status()

        self.download_url = response.headers["Location"]
        self.set_latest_file_from_download_url()

    def download(self, filename):
        print("Downloading {} to {}".format(self.download_url, filename))

        try:
            with requests.get(self.download_url, stream=True) as response:
                response.raise_for_status()
                with open(filename, "wb") as local_file:
                    for chunk in response.iter_content(chunk_size=8192):
                        if chunk:  # filter out keep-alive new chunks
                            local_file.write(chunk)
            return True
        except Exception as exception:
            print("Download failed: {}".format(exception), file=sys.stderr)
            return False

    def is_file_valid(self):
        if self.filename is None:
            return False

        if self.md5 is not None and md5file(self.filename) != self.md5:
            return None

        return True


def md5file(filename):
    md5 = hashlib.md5()

    with open(filename, "rb") as file:
        while True:
            data = file.read(65536)

            if not data:
                break

            md5.update(data)

    return md5.hexdigest()


def main():
    argument_parser = argparse.ArgumentParser(description="MineArea mods downloader")

    argument_parser.add_argument("--update", help="update all mods to their latest versions", action="store_true")

    cmd_arguments = argument_parser.parse_args()

    root_dir = os.path.dirname(os.path.realpath(__file__))
    mods_dir = os.path.join(root_dir, "source", "mods")

    if not os.path.exists(mods_dir):
        os.mkdir(mods_dir)

    mod_list_file = os.path.join(root_dir, "mods.yaml")
    update_info_file = os.path.join(root_dir, "updated-mods.txt")

    update_info_data = {
        "new": [],
        "update": []
    }

    with open(mod_list_file) as mod_list_yaml:
        mods = yaml.safe_load(mod_list_yaml)

        mods = [Mod(mods_dir, mod) for mod in mods]

        mods_to_download = []

        for mod in mods:
            if not cmd_arguments.update and mod.download_url is not None and not mod.is_file_valid():
                mods_to_download.append(mod)
                continue

            if not mod.check_url():
                continue

            if not cmd_arguments.update and mod.filename is not None:
                continue

            mod.update_latest_file()

            if mod.filename is None:
                print("New mod {} found: {}".format(mod.name, mod.latest_filename))
            else:
                current_filename = os.path.basename(mod.filename)

                if mod.latest_filename == current_filename:
                    print("No update found for {}".format(mod.name), file=sys.stderr)
                    continue

                print("Update for {} found: {} -> {}".format(mod.name, current_filename, mod.latest_filename))

            mods_to_download.append(mod)

        if not mods_to_download:
            print("Nothing to download")
            return

        print("Ready to download {} mods".format(len(mods_to_download)))
        input("Press Enter to continue...")

        for mod in mods_to_download:
            download_filepath = os.path.join(mods_dir, mod.latest_filename)
            if mod.download(download_filepath):
                old_version = mod.get_modinfo().version
                new_version = MCModInfo(download_filepath, mod.pattern).version

                if mod.filename is None:
                    update_info_data["new"].append("[{}]({})".format(mod.name, mod.url))
                elif cmd_arguments.update:
                    update_info_data["update"].append("[{}]({}): {} -> {}".format(mod.name, mod.url, old_version, new_version))

                # Remove old mod file
                if mod.filename is not None and os.path.exists(mod.filename):
                    os.remove(mod.filename)

                md5 = md5file(download_filepath)

                if cmd_arguments.update:
                    mod.md5 = md5
                else:
                    if mod.md5 is not None and mod.md5 != md5:
                        print("Warning: MD5 checksum of downloaded file does not match stored MD5 checksum!")

            elif os.path.exists(download_filepath):
                os.remove(download_filepath)

    # Dump any changes to the YAML structure (e.g. new download_url) back to the mods.yaml file
    with open(mod_list_file, "w") as mod_list_yaml:
        yaml_list = []

        for mod in mods:
            yaml_list.append(mod.to_yaml())

        yaml.dump(yaml_list, mod_list_yaml, default_flow_style=False, sort_keys=False)

    with open(update_info_file, "w") as update_info:
        if update_info_data["new"]:
            update_info.write("New mods:\n")
            for line in update_info_data["new"]:
                update_info.write("- {}\n".format(line))

        if update_info_data["update"]:
            if update_info_data["new"]:
                update_info.write("\n")

            update_info.write("Updated mods:\n")
            for line in update_info_data["update"]:
                update_info.write("- {}\n".format(line))

    print("Done")


if __name__ == "__main__":
    main()
